### ISSUE TITLE
D159 — Future detachment and strict structured child lifetime

### ISSUE THREAD

**Issue by guillermomolina** [guillermomolina/protos#623]
## Decision checkpoint

**Formal identifier:** `D159`  
**Triggered by:** AUD009-C1 / #622  
**Decision state:** `OPEN / NEEDS_USER_DECISION`

Parent: #622

Native hierarchy note: this Parent line is bootstrap input only; the native
relation is authoritative.

## Trigger

AUD009-C1 retrospectively classified the existing Core
`Future.detach()` capability as:

```text
REMOVE_NOW_RECONSIDER_LATER
```

The project owner explicitly approved the complete C1 packet, including that
classification, on 2026-09-19.

The AUD009 classification does **not** itself authorize a semantic or
implementation change. D159 owns the implementation-independent design decision
required before Core detachment can be retained, removed, relocated, or replaced.

## Exact question

Should Core keep the current `Future.detach()` operation, or should task-backed
structured child ownership become strict for the child's complete lifetime unless
a separately justified explicit lifetime mechanism is selected?

The decision must determine the observable Core contract, not merely remove one
Java method or one selector.

## Current model

Today:

```text
task-backed child created in current async task scope
    -> owned by that task scope by default

childFuture.detach()
    -> removes only that structured parent edge
    -> returns the same Future
    -> does not cancel the child
    -> does not change its eventual outcome
    -> does not re-parent it
    -> does not grant Actor-independent lifetime

non-task-backed Future.detach()
    -> no-op

terminal Future.detach()
    -> no-op
```

Detached Actor-local tasks remain subject to Actor-incarnation termination.

## AUD009-C1 evidence

At the C1 evidence baseline:

```text
PROTOS_MAIN=cfc732b154f9e1480d086f97bbc0efd06d62c37b
```

Repository search found no production use of `.detach()` under
`protos/lib` or `protos/tools`.

Current occurrences are conformance/control fixtures, primarily using detached
already-resolved/composition Futures to manufacture dependencies while testing
cancellation and cleanup behavior.

C1 identified continuing complexity from detachment in:

- the public Future protocol;
- Future detached state;
- mutable structured parent topology;
- no-op semantics for non-task-backed and terminal Futures;
- adoption/continuation ownership clauses;
- Actor-lifecycle interaction;
- tests and implementation branches explaining that detached does not mean
  Actor-independent.

The internal task parent/child ownership mechanism remains required even if the
public detachment operation is removed.

## Fixed authority

D159 must preserve unless explicitly reopened:

- D045 task-scoped structured ownership;
- Future as the public eventual-result family;
- internal Task as a non-public execution/lifetime concept;
- Actor lifetime as a separate semantic boundary;
- no public Task/scope object merely to model structured ownership;
- cooperative cancellation and ordinary Future outcome semantics;
- `Future.then` downstream-only ownership/cancellation;
- no implicit ownership change merely because a Future is returned, stored,
  wrapped, or passed through synchronous calls.

D159 must not silently redefine Actor lifetime, P/parallel isolation, I/O
producer custody, or cancellation commitment semantics.

## Mandatory candidate set

Research and compare at least:

### A — retain current Core `Future.detach()`

Keep the existing operation and current semantics.

### B — remove Core detachment with no immediate replacement

Task-backed child work remains owned by the creating asynchronous task scope
until terminal. Independent longer-lived work uses another already-existing
lifetime boundary where appropriate, such as Actors.

### C — move detachment/background capability outside Core

Investigate whether an ordinary Standard Library facility can actually express
the required lifetime semantics without privileged hidden Task authority. Reject
this candidate if it merely moves the name while retaining a special runtime
escape hatch.

### D — replace `detach` with another explicit lifetime abstraction

Consider only if current evidence shows a real requirement that strict structured
ownership plus existing Actor/runtime boundaries cannot meet. Do not invent a
public Task/scope/background-job institution merely for familiarity.

A credible additional candidate discovered by comparative research must be
included rather than forced into one of these labels.

## Required comparative research

Follow current `AGENTS.md` and `spec/AGENTS.md`.

The Dxxx packet must survey at least five credible systems spanning at least
three materially different concurrency/lifetime approaches and compare the
underlying ownership semantics rather than API spelling.

Relevant families should include, where applicable:

- structured-concurrency systems with strict lexical/task ownership;
- systems with explicit detached/unstructured tasks;
- actor/process lifetime models;
- futures/promises where result handles and execution ownership are separate.

The comparison must cover cancellation, failure, cleanup, lifetime escape,
resource retention, Actor/process interaction, memory scaling, implementation
freedom and migration consequences.

## Required falsification

At minimum test:

1. whether strict ownership makes legitimate daemon/background work impossible
   without forcing Actors;
2. whether Actor creation is too heavyweight or semantically wrong as the only
   independent-lifetime alternative;
3. whether removing detachment can create accidental unbounded waits at owner
   completion;
4. whether retaining detachment enables leaked/unowned work that structured
   concurrency was meant to prevent;
5. whether a library-only detachment facility can be honest without hidden Core
   authority;
6. whether another public lifetime abstraction would add more institutions than
   it removes;
7. whether future reintroduction after removal would require foundational
   ownership redesign or only a bounded new operation/abstraction.

## Required decision packet

Before requesting project-owner selection, D159 must provide the complete
current-AGENTS decision packet, including:

- prior-art survey;
- full candidate set;
- 12-dimension scoring with confidence;
- strongest counterexamples and failure modes;
- pay-for-what-you-need / grow-as-you-need analysis;
- concrete cost of deferral;
- scalability and future-scenario stress;
- compatibility/migration consequences;
- recommendation;
- strongest argument against the recommendation;
- intentionally deferred questions.

No candidate is selected by opening D159.

## Downstream routing

If the selected D159 outcome changes observable Core semantics:

1. publish the durable D159 decision record;
2. update the applicable normative concurrency/callable specifications and
   `spec/PROTOS_SPEC_CHANGELOG.md`;
3. route implementation/test removal through the appropriate implementation
   owner;
4. remove obsolete documentation and conformance expectations consistently.

No implementation work is authorized until D159 is ratified.

## Approval state

```text
AUD009_C1_CLASSIFICATION_APPROVED=YES
D159_CANDIDATE_SELECTED=NO
D159_STATUS=NEEDS_USER_DECISION
```

Formal child relationship: AUD009-C1 / #622.

**Comment by guillermomolina** [comment]
## D159 ratification closure

The project owner explicitly approved **Candidate B** in the active interaction on 2026-09-19.

Selected boundary:

```text
Future.detach()                              REMOVE_NOW_RECONSIDER_LATER
task-backed child ownership                  STRICT_COMPLETE_LIFETIME
background / independent Actor-local work    DEFERRED
public Task / scope                          REMAINS ABSENT
```

The approval is not based merely on current non-use. D159 distinguishes the D158 case: `Future.detach()` is a public post-creation mutation of structured lifetime ownership through an eventual-result handle. Removing it establishes the stronger invariant that task-backed child ownership remains stable until the child is terminal.

GITHUB021 invariant/delta review passes: D045 task-scoped ownership, Future identity/outcomes, hidden Task, Actor lifetime, cooperative cancellation, `Future.then` downstream behavior, P isolation and I/O custody remain preserved/unaffected.

Durable ratification was validated with `git diff --check`, published, and re-read at the exact revision.

Implementation is routed separately to **I055 / #656**. I055 is `status:ready`, `priority:p1`, assigned to the repository owner, and consumes this exact ratification. It does not select a replacement background-work API.

```text
FORMAL_IDENTIFIER_UNIQUE=PASS
FAMILY=family:D PASS
CANONICAL_STATUS=status:completed PASS_ON_CLOSURE
ASSIGNEE_INVARIANT=PASS
NATIVE_PARENT=#622 PASS
ANCESTOR_PRIORITY=#522 priority:p1
EFFECTIVE_PRIORITY=RESOLVED
PROJECT_ROUTING=PASS
DECISION_APPROVAL_PROVENANCE=PASS
DECISION_INVARIANT_CONSISTENCY=PASS

PROTOS_REVISION=ecf563ed01275929d5b85330e8e6259cc85d73d8
PROJECT_RECORD_REVISION=3a40a27b4784945e47ca6e4e9d735447a2326f8c
PROJECT_RECORD=docs/project/decisions/language/D159_FUTURE_DETACHMENT_AND_STRICT_STRUCTURED_CHILD_LIFETIME.md
REQUIRED_DURABLE_PUBLICATION=PASS

IMPLEMENTATION_OWNER=I055/#656
IMPLEMENTATION_OWNER_STATE=status:ready
IMPLEMENTATION_OWNER_PRIORITY=priority:p1
CROSS_REFERENCES=PASS
```

D159 is ratified; normative/runtime/test changes remain pending under I055.
